File: 1125 (Smallest Sufficient Team)/cpp/Solution.cpp
class Solution {
public:
    vector<int> smallestSufficientTeam(vector<string>& req_skills, vector<vector<string>>& people) {
        unordered_map<string,int> skill_map;
        for(int i= 0;i<req_skills.size();i++){
            skill_map[req_skills[i]] = i;
        }
        unordered_map<int,vector<int>> dp;
        dp[0] = {};

        for(int index =0;index<people.size();index++){
            int person_bits = 0;
            for(int i = 0;i<people[index].size();i++){
                person_bits |= (1<<skill_map[people[index][i]]);
            }

            vector<int> tasks;
            for(const auto& entry: dp){
                tasks.push_back(entry.first);
            }

            for(int i = 0;i<tasks.size();i++){
                int new_tasks = tasks[i]|person_bits;
                if (dp.find(new_tasks) == dp.end() || (dp[new_tasks].size()>dp[tasks[i]].size()+1)){
                    dp[new_tasks] = dp[tasks[i]];
                    dp[new_tasks].push_back(index);
                }
            }
        }
        return dp[(1<<(req_skills.size()))-1];
    }
};

File: 1125 (Smallest Sufficient Team)/python/Solution.py
class Solution:
    def smallestSufficientTeam(self, req_skills: List[str], people: List[List[str]]) -> List[int]:

        skill_map = {skill: index for index,skill in enumerate(req_skills)}
        dp = {0: []} # 001 , 010, 011 = [1,3], 100, 111 = [1,3,9]

        for index in range(len(people)):
            person_bits =  0
            for skill in people[index]:
                person_bits |= (1<<skill_map[skill])
            
            tasks = list(dp.keys())
            for i in range(len(tasks)):
                new_task = tasks[i] | person_bits # 011 | 100 = 111

                if new_task not in dp or len(dp[new_task]) > len(dp[tasks[i]])+1:
                    dp[new_task] = dp[tasks[i]]+[index]
            
        # 111
        # 3 => 1*2^(3) = 1000 -1 = 111
        return dp[(1<<len(req_skills))-1]
